hudi-client/hudi-client-common/src/main/java/org/apache/hudi/optimize/HilbertCurve.java

```diff
@@ -0,0 +1,321 @@
+   * @return index (nonnegative {@link BigInteger})
+   * @throws IllegalArgumentException
+   *             if length of point array is not equal to the number of
+   *             dimensions.
+   */
+  public BigInteger index(long... point) {
+    if (point.length != dimensions) {
+      throw new IllegalArgumentException(String.format("length of point array must equal to the number of dimensions"));
+    }
+    return toIndex(transposedIndex(bits, point));
+  }
+
+  public byte[] indexBytes(long... point) {
+    if (point.length != dimensions) {
+      throw new IllegalArgumentException(String.format("length of point array must equal to the number of dimensions"));
+    }
+    return toIndexBytes(transposedIndex(bits, point));
+  }
+
+  /**
+   * Converts a {@link BigInteger} index (distance along the Hilbert Curve from 0)
+   * to a point of dimensions defined in the constructor of {@code this}.
+   *
+   * @param index
+   *            index along the Hilbert Curve from 0. Maximum value 2 <sup>bits *
+   *            dimensions</sup>-1.
+   * @return array of longs being the point
+   * @throws NullPointerException
+   *             if index is null
+   * @throws IllegalArgumentException
+   *             if index is negative
+   */
+  public long[] point(BigInteger index) {
+    if (index == null) {
+      throw new NullPointerException("index must not be null");
+    }
+    if (index.signum() == -1) {
+      throw new IllegalArgumentException("index cannot be negative");
+    }
+    return transposedIndexToPoint(bits, transpose(index));
+  }
+
+  public void point(BigInteger index, long[] x) {
+    if (index == null) {
+      throw new NullPointerException("index must not be null");
+    }
+    if (index.signum() == -1) {
+      throw new IllegalArgumentException("index cannot be negative");
+    }
+    Arrays.fill(x, 0);
+    transpose(index, x);
+    transposedIndexToPoint(bits, x);
+  }
+
+  public void point(long i, long[] x) {
+    point(BigInteger.valueOf(i), x);
+  }
+
+  /**
+   * Converts a {@code long} index (distance along the Hilbert Curve from 0) to a
+   * point of dimensions defined in the constructor of {@code this}.
+   *
+   * @param index
+   *            index along the Hilbert Curve from 0. Maximum value 2
+   *            <sup>bits+1</sup>-1.
+   * @return array of longs being the point
+   * @throws IllegalArgumentException
+   *             if index is negative
+   */
+  public long[] point(long index) {
+    return point(BigInteger.valueOf(index));
+  }
+
+  /**
+   * Returns the transposed representation of the Hilbert curve index.
+   *
+   * <p>
+   * The Hilbert index is expressed internally as an array of transposed bits.
+   *
+   * <pre>
+   Example: 5 bits for each of n=3 coordinates.
+   15-bit Hilbert integer = A B C D E F G H I J K L M N O is stored
+   as its Transpose                        ^
+   X[0] = A D G J M                    X[2]|  7
+   X[1] = B E H K N        &lt;-------&gt;       | /X[1]
```

Comment:
This seems off

Reply:
Thank you for your review。 may be i miss somethings， could you point out the problem more details thanks。

Reply:
@alexeykudinkin can you please expand. so @xiarixiaoyao can address/

Reply:
Sorry, for not being elaborate enough -- i was referring to escaped symbols `&lt;-------&gt;`

Reply:
oh, sorry  let me fixed it.